[DO NOT MERGE] Add new documentation for variable sets (#1950)
hashicorp/terraform-website: content/source/docs/cloud/workspaces/variables.html.md

---
layout: "cloud"
page_title: "Workspace Variables - Terraform Cloud and Terraform Enterprise"
description: "Terraform Cloud workspace variables let you customize configurations, modify Terraform's behavior, and store information like provider credentials."

---

# Variables

> -> **Note:** Variable sets are in beta.

Reviewer: ```suggestion
~> **Note:** Variable sets are in beta.
```

Author: done - good call.

You must have [`read variables` permission](/docs/cloud/users-teams-organizations/permissions.html#general-workspace-permissions) to view the variables for a particular workspace and to view the variable sets in your organization. Once you have the proper [read and write variables permissions](/docs/cloud/users-teams-organizations/permissions.html#general-workspace-permissions), you can create and edit both workspace-specific variables and variable sets through:

- The [Terraform Cloud UI](/docs/cloud/workspaces/managing-variables.html).
- The Variables API for [workspace-specific variables](/docs/cloud/api/workspace-variables.html) and [variable sets](/docs/cloud/api/variable-sets.html).
- The `tfe` provider [`tfe_variable`](https://registry.terraform.io/providers/hashicorp/tfe/latest/docs/resources/variable) resource, which can be more convenient for large numbers of complex variables.

[permissions-citation]: #intentionally-unused---keep-for-maintainers

## Types

You can create both Terraform variables and environment variables in Terraform Cloud.

> > **Hands-on:** The [Create Infrastructure tutorial](https://learn.hashicorp.com/tutorials/terraform/cloud-workspace-configure?in=terraform/cloud-get-started) on HashiCorp Learn shows how to set both Terraform variables and provider credential environment variables.

Reviewer: @ritsok modified this tutorial flow a bit, so I don't think this is accurate. Can we update this?

https://learn-git-addvarsettfcgs-hashicorp.vercel.app/tutorials/terraform/cloud-create-variable-set

Author: yes! thank you - updated!

### Terraform Variables

> Terraform variables refer to [input variables](/docs/language/values/variables.html) that define parameters without hardcoding them into the configuration. For example, you could create variables that let users specify the number and type of Amazon Web Services EC2 instances they want to provision with a Terraform module. If a required input variable is missing, Terraform plans in the workspace will fail and print an explanation in the log.

Author: good idea!!! I pulled an example from learn (yay learn!) that I'm pushing up now.

> Terraform Cloud passes input variables to Terraform by writing a `terraform.tfvars` file, meaning that Terraform Cloud overwrites any `terraform.tfvars` file you check into version control. You should not check in `terraform.tfvars` files even when running Terraform solely on the command line.

Reviewer 2: The other thing I'm wondering about is whether or not readers know enough about TFC at this point to understand the statement "Terraform Cloud passes input variables to Terraform". The environment variables section introduces some of the architecture that might help users understand this statement better, so it may be better to explain that first?

> Terraform Cloud passes input variables to Terraform by writing a `terraform.tfvars` file, meaning that Terraform Cloud overwrites any `terraform.tfvars` file you check into version control. You should not check in `terraform.tfvars` files even when running Terraform solely on the command line.

Author: Yeah you know what - this was very confusing. I clarified with engineering and am pushing up a fix that I think makes this clearer. That's also an interesting point about knowing about TFC and how it works. I switched the order (environment variables) now come first in this section and I also modified the way it's written to hopefully make it a bit clearer as well.

> You must define input variables manually; Terraform Cloud does not automatically discover variable names from Terraform code. However, workspaces that use Terraform 0.10.0 or later allow you to commit any number of [`*.auto.tfvars` files](/docs/language/values/variables.html#variable-files) to provide default Terraform input variables. If any variables from these files have the same key as existing variables in the Terraform Cloud workspace, the variables applied to the workspace overwrite variables from the files.

Reviewer: "Terraform Cloud does not automatically discover variable names from Terraform code" is this true? I thought TFC would use the default value when you declare the variables in your config

Author: this is confusingly written - let me work this out and push up a fix. I need to confirm with eng :-)

Author: Okay I'm pushing up a big fix to this that I think makes it a lot clearer. Thank you for calling this out - this happened as I was reworking the page (pulling content from various places) and it didn't mesh quite right. I'm hoping what I'm pushing up is better!

### Environment Variables

Terraform Cloud performs Terraform runs on disposable Linux worker VMs using a POSIX-compatible shell. Before running Terraform operations, Terraform Cloud uses the `export` command to populate the shell with environment variables.

Environment variables can store provider credentials and other data. Refer to your provider's Terraform Registry documentation for a full list of supported shell environment variables (e.g., authentication variables for [AWS](https://registry.terraform.io/providers/hashicorp/aws/latest/docs#environment-variables), [Google Cloud Platform](https://registry.terraform.io/providers/hashicorp/google/latest/docs/guides/getting_started#adding-credentials), and [Azure](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs#argument-reference)). Environment variables can also [modify Terraform's behavior](/docs/cli/config/environment-variables.html). For example, `TF_LOG` enables detailed logs for debugging.

> #### Parallelism

Reviewer 2: This section seems a little bit in the weeds compared to the rest of the content on this page. Not sure if there's anywhere better to put it though.

Author: yeah this is a relic of the old content. I wanted to combine things - there used to be a "Special Environment Variables" section, which I thought was a bit weird. I don't love this, but....I think I'll keep it for now. But I agree - it is definitely very specific given the rest of the content on the page.

> You can use the `TFE_PARALLELISM` environment variable when your infrastructure providers produce errors on concurrent operations or use non-standard rate limiting. The `TFE_PARALLELISM` variable sets the  `-parallelism=<N>` flag for  `terraform plan` and `terraform apply`  ([more about `parallelism`](/docs/internals/graph.html#walking-the-graph)). Valid values are between 1 and 256, inclusive, and the default is `10`. Terraform Cloud Agents do not support `TFE_PARALLELISM`, but you can specify flags as environment variables directly via [TF_CLI_ARGS](/docs/cli/config/environment-variables.html#tf-cli-args). In these cases, use `TF_CLI_ARGS="parallelism=<N>"` instead.

Reviewer 2: ```suggestion
You can use the `TFE_PARALLELISM` environment variable when your infrastructure providers produce errors on concurrent operations or use non-standard rate limiting. The `TFE_PARALLELISM` variable sets the  `-parallelism=<N>` flag for  `terraform plan` and `terraform apply`  ([more about `parallelism`](/docs/internals/graph.html#walking-the-graph)). Valid values are between 1 and 256, inclusive, and the default is `10`. Terraform Cloud Agents do not support `TFE_PARALLELISM`, but you can specify flags as environment variables directly via [`TF_CLI_ARGS`](/docs/cli/config/environment-variables.html#tf-cli-args). In these cases, use `TF_CLI_ARGS="parallelism=<N>"` instead.
```

> ~> **Warning:** We recommend talking to HashiCorp support before setting `TFE_PARALLELISM`.

Reviewer: ```suggestion
!> **Warning:** We recommend talking to HashiCorp support before setting `TFE_PARALLELISM`.
```

> ## Scope

Reviewer 2: I feel like this section is kind of a repeat of what we talked about in the page intro. It might be nicer to provide a reference table here with ways to set variables and their scopes, so that people can quickly find the scope they want the variable to have and match it to a way to set the variable?

Author: Judith - this is an amazing idea. I'm going to implement it now.

You can [set workspace-specific variables](/docs/cloud/workspaces/managing-variables.html#workspace-specific-variables) inside each workspace or [create variable sets](/docs/cloud/workspaces/managing-variables.html#variable-sets) to reuse the same variables across multiple workspaces. For example, you could define a variable set of provider credentials and automatically apply it to several workspaces, rather than manually defining credential variables in each. Changes to variable sets instantly apply to all appropriate workspaces, saving time and reducing errors from manual updates.

> You can create a global variable set by applying it to all current and future workspaces. Terraform Cloud does not allow global variable sets to contain conflicting variables with same type and the same key. When non-global variable sets contain conflicting variables and you apply them to the same workspace, the variable set applied most recently takes precedence. Refer to the [precedence documentation](#precedence) for more details.

Reviewer 3: ```suggestion
You can create a global variable set by selecting the `Apply to all workspaces in this organization` scope in the variable set's settings. Terraform Cloud does not allow global variable sets to contain conflicting variables with same type and the same key. When non-global variable sets contain conflicting variables and you apply them to the same workspace, the variable set applied most recently takes precedence. Refer to the [precedence documentation](#precedence) for more details.
```

## Precedence

> > **Hands On:** The [Manage Multiple Variable Sets in Terraform Cloud](link) tutorial on HashiCorp Learn shows how to manage multiple variable sets and demonstrates variable precedence.

Reviewer: ```suggestion
> **Hands On:** The [Manage Multiple Variable Sets in Terraform Cloud](https://learn.hashicorp.com/tutorials/terraform/manage-variable-sets) tutorial on HashiCorp Learn shows how to manage multiple variable sets and demonstrates variable precedence.
```

> > **Hands On:** The [Manage Multiple Variable Sets in Terraform Cloud](link) tutorial on HashiCorp Learn shows how to manage multiple variable sets and demonstrates variable precedence.

Reviewer 3: ```suggestion
> **Hands On:** The [Manage Multiple Variable Sets in Terraform Cloud](https://learn.hashicorp.com/tutorials/terraform/manage-variable-sets) tutorial on HashiCorp Learn shows how to manage multiple variable sets and demonstrates variable precedence.
```

If variables from `.auto.tfvars` files have the same key as existing variables in the Terraform Cloud workspace, the variables applied to the workspace overwrite variables from the files.

> There are two additional cases when a workspace may have conflicting variables with the same key. You may use workspace-specific variables to [overwrite variables from variable sets](/docs/cloud/workspaces/managing-variables.html#overwrite-variable-sets) or there may be two variables from different variable sets with the same key.

Reviewer 3: this section feels unclear to me. maybe we can organize the different overwrite scenarios under subheadings? something like
- conflicts between workspace-specific variables and variable sets
- conflicts between multiple variable sets

Reviewer 2: I think usually when I look for a precedence section I expect a list of all the ways you can set the thing from highest to lowest precedence.

Author: okay I'm pushing up a big change here - you're right. This section is not working. I'm hoping what I'm pushing up is better!

Workspace-specific variables always overwrite variables from variable sets that have the same key.

If different variable sets contain variables of the same type with the same key, Terraform Cloud prioritizes them first based on the variable set scope and then by the date that each variable set was applied to the workspace.

- **Scope:** Variable sets that apply to only a subset of workspaces take precedence over global variable sets that are applied to all workspaces within an organization.

> - **Date Applied:** If non-global variable sets have conflicting variables, Terraform cloud prioritizes the variable set that was applied most recently. For example, if you apply Variable Set A and then apply Variable Set B to the same workspace a week later, Terraform Cloud will use any conflicting variables from Variable Set B. This will be the case regardless of which variable set has been edited most recently; Terraform Cloud only considers the date that each variable set is applied to the workspace when determining precedence.

Reviewer 3: ```suggestion
- **Date Applied:** If non-global variable sets have conflicting variables, Terraform Cloud uses values from the variable set that was applied most recently. For example, if you apply Variable Set A and then apply Variable Set B to the same workspace a week later, Terraform Cloud will use any conflicting variables from Variable Set B. This will be the case regardless of which variable set has been edited most recently; Terraform Cloud only considers the date that each variable set is applied to the workspace when determining precedence.
```

Terraform Cloud displays a message in the UI when it overwrites conflicting variables.

> ![A Terraform Cloud UI message explaining which variables are overwritten](link)

Reviewer 2: This screenshot doesn't render yet :)

Author: i know! that is coming - promise!

### Precedence Example

Consider an example workspace that has:

**Workspace-specific variables**

- ACCESS_KEY = g
- ACCESS_ID = q

> - VAR1 = h

Reviewer: might be easier to parse as a table

Reviewer 2: Also make sure to code format these

Author: done!

**Variable Set A (applied 10/4)**

- KEY1 = x
- VAR1 = y

**Global Variable Set B (Applied 10/20)**

- KEY1 = z
- VAR2 = a

When you trigger a run, Terraform Cloud applies the following variables:

> - **Workspace-Specific:** ACCESS_KEY, ACCESS_ID, and VAR1. That means VAR1 equals h for this run, overwriting the value in Variable Set A.

Reviewer: might be easier to parse if the variable names and values were codefenced ``

Reviewer: ```suggestion
- **Workspace-Specific:** `ACCESS_KEY`, `ACCESS_ID`, and `VAR1`. That means `VAR1` equals `h` for this run, overwriting the value in Variable Set A.
```

- **Variable Set A:** KEY1. That means KEY1 equals x for this run, overwriting the value in Variable Set B.

- **Variable Set B:** VAR2. This is a global variable set, so Variable Set A takes precedence regardless of when it was applied.

> ![An example scenario demonstrating variable precedence in Terraform Cloud](image)

Reviewer 2: Remember to update with an image

Author: yes! will do!

~> **Note:** Variable sets are in beta.

You must have [`read variables` permission](/docs/cloud/users-teams-organizations/permissions.html#general-workspace-permissions) to view the variables for a particular workspace and to view the variable sets in your organization. Once you have the proper [read and write variables permissions](/docs/cloud/users-teams-organizations/permissions.html#general-workspace-permissions), you can create and edit both workspace-specific variables and variable sets through:

- The [Terraform Cloud UI](/docs/cloud/workspaces/managing-variables.html).
- The Variables API for [workspace-specific variables](/docs/cloud/api/workspace-variables.html) and [variable sets](/docs/cloud/api/variable-sets.html).
- The `tfe` provider [`tfe_variable`](https://registry.terraform.io/providers/hashicorp/tfe/latest/docs/resources/variable) resource, which can be more convenient for large numbers of complex variables.

[permissions-citation]: #intentionally-unused---keep-for-maintainers

## Types

You can create both  environment variables and Terraform variables in Terraform Cloud.

> > **Hands-on:** The [Create Infrastructure tutorial](https://learn.hashicorp.com/tutorials/terraform/cloud-workspace-configure?in=terraform/cloud-get-started) on HashiCorp Learn shows how to set Terraform variables and and the [Create and Use a Variable Set](https://learn.hashicorp.com/tutorials/terraform/cloud-create-variable-set?in=terraform/cloud-get-started) tutorial shows how to set provider credential environment variables.

Reviewer 2: ```suggestion
> **Hands-on:** Try the [Create and Use a Variable Set](https://learn.hashicorp.com/tutorials/terraform/cloud-create-variable-set?in=terraform/cloud-get-started) tutorial and [Create Infrastructure tutorial](https://learn.hashicorp.com/tutorials/terraform/cloud-workspace-configure?in=terraform/cloud-get-started) on HashiCorp Learn to set environment and Terraform variables in Terraform Cloud.
```

!> **Warning:** We recommend talking to HashiCorp support before setting `TFE_PARALLELISM`.

### Terraform Variables

Terraform variables refer to [input variables](/docs/language/values/variables.html) that define parameters without hardcoding them into the configuration. For example, you could create variables that let users specify the number and type of Amazon Web Services EC2 instances they want to provision with a Terraform module.

``` hcl
variable "instance_count" {
  description = "Number of instances to provision."
  type        = number
  default     = 2
}
```

You can then reference this variable in your configuration.

``` hcl
module "ec2_instances" {
  source = "./modules/aws-instance"

> -  instance_count = 2

Author: fixed! sorry about that

+  instance_count = var.instance_count
  ## ...
}
```

If a required input variable is missing, Terraform plans in the workspace will fail and print an explanation in the log.

#### Loading Variables from Files

You can provide default variable values by committing any number of [files ending in `.tfvars`](/docs/language/values/variables.html#variable-files) to workspaces that use Terraform 0.10.0 or later. If any variables from these files have the same key as existing variables in the Terraform Cloud workspace, the variables applied to the workspace overwrite variables from the files.

> One exception is `terraform.tfvars` files. Terraform Cloud creates a `terraform.tfvars` file to pass a workspace's input variables to Terraform during runs. This means that Terraform Cloud will overwrite any `terraform.tfvars` file you check into version control. You can check in other types of `.tfvars` files, but do not check in `terraform.tfvars` files even when running Terraform solely on the command line.

Author: So I went to the expert this morning - @chrisarcand and we worked on a fix for this together! I am confident that it's correct now and also not as confusing. In short, *auto.tfvars is the only type of file you can use right now, so we're very explicit about that. And we took out the extraneous stuff about terraform.tfvars because the user doesn't really need to know it - they just need to know what does work and that they can't use any other type of file.

> ~> **Note:** Terraform Cloud can infer default variable values, but it does not automatically display those variable names in the workspace UI.

Reviewer 2: ```suggestion
~> **Note:** When Terraform Cloud loads variable values from files it does not automatically display those variable names in the workspace UI.
```

## Scope

Each environment and Terraform variable can have one of the following scopes:

| Scope | Description| Resources |
|-------|------------|-----------|
|Workspace-Specific | Apply to a single workspace | [Create Workspace-Specific Variables](/docs/cloud/workspaces/managing-variables.html#workspace-specific-variables), [Loading Variables from Files](#loading-variables-from-files), [Workspace-Specific Variables API](/docs/cloud/api/workspace-variables.html).|
|Variable Set | Apply to multiple workspaces within the same organization. | [Create Variable Sets](/docs/cloud/workspaces/managing-variables.html#variable-sets) and [Variable Sets API](/docs/cloud/api/variable-sets.html)|
|Global Variable Set | Automatically applied to all current and future workspaces within an organization. | [Create Variable Sets](/docs/cloud/workspaces/managing-variables.html#variable-sets) and [Variable Sets API](/docs/cloud/api/variable-sets.html)|

## Precedence

> > **Hands On:** The [Manage Multiple Variable Sets in Terraform Cloud](https://learn.hashicorp.com/tutorials/terraform/manage-variable-sets) tutorial on HashiCorp Learn shows how to manage multiple variable sets and demonstrates variable precedence.

Reviewer 3: ```suggestion
> **Hands On:** The [Manage Multiple Variable Sets in Terraform Cloud](https://learn.hashicorp.com/tutorials/terraform/cloud-multiple-variable-sets) tutorial on HashiCorp Learn shows how to manage multiple variable sets and demonstrates variable precedence.
```

There may be cases when a workspace contains conflicting variables of the same type with the same key. Terraform Cloud marks overwritten variables in the UI.

> ![The Terraform Cloud UI indicates which variables are overwritten](/docs/cloud/workspaces/images/ui-overwritten-variables.png)

Reviewer 2: Is this the right resolution?

Author: it actually was, but it looked weird I think because of the way I tried to zoom into a tiny part of the screen - I redid this one and it looks much better now!

### 3. Global Variable Sets

Non-global variable sets always take precedence over global variable sets that are applied to all workspaces within an organization. This is true regardless of when the global variable set was applied.

### 4. Variable Files

When variables from [committed `.auto.tfvars` files](#loading-variables-from-files) have the same key as existing variables in the Terraform Cloud workspace, the variables applied to the workspace overwrite variables from the files.

### Precedence Example

Consider an example workspace that has:

| Workspace-Specific Variables | Variable Set A | Variable Set B (Global)|
| -----------------------------|----------------|----------------|
|   | Applied 10/4 | Applied 10/20 |

> | `ACCESS_KEY` = `g`, `ACCESS_ID` = `q`, `VAR1` = `h` | `KEY1` = `x`, `VAR1` = `y` | `KEY1` = `z`, `VAR2` = `a` |

Reviewer 2: Can we make the keys the row titles in the table, and give each variable its own row? That would make it easier to tell which variables are overlapping between sources.

Author: yes! good idea - I did that just now and it's much better!
